src/typing/env_js.ml

```diff
@@ -661,24 +771,21 @@ let merge_env =
           (string_of_reason reason)
           name)
       (* global lookups may leave new entries in one or both child envs *)
-      | None, child1, child2 when env0 = [] ->
+      | None, _, _ when env0_ = [] ->
         (* ...in which case we can forget them *)
         ()
-      (* otherwise, non-refinement uneven distributions are asserts. *)
-      (* TODO:
-         Asserting on all uneven distributions is too strict for now.
-         let_env can interact with path-dependent refinement to leave
-         residual entries in child contexts, when catch blocks combine
-         conditionals and early returns.
-         So for now we need to let extra child entries pass, though we
-         should still assert on extra parent entries.
-         Once let/const/LexScope is in place, we can reimplement let_env
-         in terms of those constructs instead, and the full assert can
-         be reinstated.
-       *)
-      | None, _, _ ->
+      (* changeset entry exists only in lex scope *)
+      | None, None, Some e when is_lex scope2 && Entry.is_lex e ->
+        ()
+      | None, Some e, None when is_lex scope1 && Entry.is_lex e ->
         ()
-      | Some _ as orig, child1, child2 ->
+      (* walk through uneven lex scopes *)
```

Comment:
Your call. Personally I'd do the pair cases explicitly rather than relying on the recursion just to convey intent, but I don't want to peck your commit to death. :) Do whatever feels right - if I see a way to matricize the patterns in here a bit more tightly that's worth the churn, I'll fly it in after your PR is merged.